### ISSUE TITLE
loss 개선

### ISSUE THREAD

**Issue by krikit** [ecarg/grace#8]
현재는 loss를 계산하는 함수로 CrossEntropyLoss를 사용합니다.
출력 레이블 O에 대해 loss가 과도하게 부여되는 면이 있으므로 이를 개선하는 loss 함수를 새로 구현합니다.

**Comment by kobikun** [comment]
* loss 함수에 loss term을 추가해서 이를 조정하면 성능이 향상된다는 글이 있네요

```간단한 아이디어로 꽤 많은 변화를 끌어낸 연구에 대한 포스트입니다. loss function으로 흔히 사용되는 cross entropy에 center loss term 을 추가하고 비율을(lambda) 조정해 가면서 실험을 한 것인데, 매우 많은 실험에서 결과가 더 좋아지고, 이 리뷰를 쓴 친구의 실험에서도 재현이 되네요. Center loss term의 아이디어는 네트워크가 gradient descent 를 통해서 업데이트 됨에 따라 각 클래스의 center 가 변하는데 이를 업데이트할 필요가 있다는 간단한 생각에서 시작이 되었고, 이를 통해 오버피팅도 덜 일어나게 할 수 있을 것이라는 어찌 보면 매우 단순한 생각으로 시작된 실험의 결과입니다. 현재 사용하고 있는 방법이 틀릴 수 있다고 생각하고 의심하는 것이 좋은 연구의 첩경이라는 것을 보여주는 사례 중 하나가 아닐까 싶네요.```

https://medium.com/mlreview/experiments-with-a-new-loss-term-added-to-the-standard-cross-entropy-85b080c42446

**Comment by krikit** [comment]
* 기존과의 비교를 위해 우선 현재 상태를 기록해 둡니다.
* 음절의 태그 정확도 및 그때의 loss입니다.

1) FNN
* loss: 0.041392755
* accuracy: 93.99001464

2) CNN
* loss: 0
* accuracy: 95.60794324

___log_softmax를 두번 취한 오류가 있어 값을 정정합니다.___

**Comment by krikit** [comment]
첫번째 시도로 forward() 메소드에서 마지막 output layer 계산을 완료한 후 최대 값이 O일 경우 softmax 값으로 loss를 계산하지 않고, one hot으로 변환하여 loss를 계산하도록 합니다.

가령 output layer가 B, I, O 세가지 카테고리만 있다고 할 때, forward() 메소드의 최종 출력 값이 [0.1, 0.2, 0.7]이라면, 이 값을 training 시 [0, 0, 1]으로 one hot으로 변환해서 loss를 계산하도록 합니다.

그러면 정답이 O인 경우, 즉 [0, 0, 1]인 경우 loss가 0이 되어 gradient가 0이되고 backward() 타임에 파라미터 업데이트를 하지 않게 됩니다.

그러나 정답이 I인 경우, 즉 [0, 1, 0]인 경우는 반대로 loss가 기존 대비 큰 값이 되어 backward() 타임에 파라미터를 크게 업데이트 하게 됩니다.

___NLL loss를 이용해 여러 시도를 해 봤으나 back propagation 알고리즘이 제대로 작동하지 않아 실패했습니다.___

**Comment by krikit** [comment]
두번째 시도로 loss 함수에 weight 생성자 파라미터를 통해 카테고리 별로 가중치를 줄 수 있습니다.
[CrossEntropyLoss](http://pytorch.org/docs/master/nn.html#torch.nn.CrossEntropyLoss)

O 태그에 대해서만 가중치를 0.1로 낮춘 것과 10으로 높이는 실험을 진행했고 결과는 아래와 같습니다.

### 0.1로 낮춘 결과

1. FNN
* loss: 0.500896871089935
* accuracy: 92.4997184579

2. CNN
* loss: 0.033473215997219
* accuracy: 94.6619617853523

### 10으로 높인 결과

1. FNN
* loss: 0.033191580325365
* accuracy: 93.4532077029918

2. CNN (진행중, 760k-th iteration)
* loss: 0.0079787578433752
* accuracy: 94.9059649386238

___weight를 조정한 결과도 그대로 둔 것에 비해 낮게 나옵니다.___
[loss_acc.xlsx](https://github.com/ecarg/grace/files/1294804/loss_acc.xlsx)

**Comment by krikit** [comment]
develop 브랜치에 머지하지 않고 이슈를 그냥 닫습니다.

단, CrossEntropyLoss를 사용하므로 모델의 forward 마지막에 log_softmax를 취하는 부분은 제거하는 것이 좋을 것 같습니다.
